packages/@ember/array/index.ts

```diff
@@ -1935,6 +1942,127 @@ NativeArray.keys().forEach((methodName) => {
 
 NativeArray = NativeArray.without(...ignore);
 
+const NATIVE_ARRAY_METHODS = Object.freeze(NativeArray.keys());
+
+function getCP(decorator: Function) {
+  let desc = descriptorForDecorator(decorator);
+  assert('[BUG] Expected descriptor with getter', desc instanceof ComputedProperty);
+  return desc;
+}
+
+// NOTE: This code borrows heavily from https://github.com/tracked-tools/tracked-built-ins/blob/master/addon/src/-private/array.ts
+
+// Unfortunately, TypeScript's ability to do inference *or* type-checking in a
+// `Proxy`'s body is very limited, so we have to use a number of casts `as any`
+// to make the internal accesses work. The type safety of these is guaranteed at
+// the *call site* instead of within the body: you cannot do `Array.blah` in TS,
+// and it will blow up in JS in exactly the same way, so it is safe to assume
+// that properties within the getter have the correct type in TS.
+class EmberAProxy<T = unknown> {
+  /**
+   * Creates an array from an iterable object.
+   * @param iterable An iterable object to convert to an array.
+   */
+  static from<T>(iterable: Iterable<T> | ArrayLike<T>): EmberAProxy<T> {
+    return new EmberAProxy(Array.from(iterable));
+  }
+
+  static of<T>(...arr: T[]): EmberAProxy<T> {
+    return new EmberAProxy(arr);
+  }
+
+  constructor(arr: T[] = []) {
+    let self = this;
+
+    let proxy = new Proxy(arr, {
+      get(target, prop /*, _receiver */) {
+        // These nonEnumerableComputed properties need special handling
+        if (prop === '[]') {
```

Comment:
Ideally we'd have something more generic, but we aren't going to be adding stuff to EmberArray going forwards so this should be fine.